File: internal/handler/members/members_presentation.go
package members

import (
	"github.com/gin-gonic/gin"
	"gitlab.com/gitlab-org/gitlab-redux/internal/service"
)

// MembersPresentation handles the presentation of members
type MembersPresentation struct {
	membersPreloader *service.MembersPreloader
	presenterFactory *service.PresenterFactory
	logger           *service.Logger
}

// NewMembersPresentation creates a new instance of MembersPresentation
func NewMembersPresentation(
	membersPreloader *service.MembersPreloader,
	presenterFactory *service.PresenterFactory,
	logger *service.Logger,
) *MembersPresentation {
	return &MembersPresentation{
		membersPreloader: membersPreloader,
		presenterFactory: presenterFactory,
		logger:           logger,
	}
}

// PresentMembers presents a list of members
func (m *MembersPresentation) PresentMembers(ctx *gin.Context, members []*service.Member) ([]*service.MemberPresenter, error) {
	// Preload associations for the members
	if err := m.preloadAssociations(members); err != nil {
		return nil, err
	}

	// Get current user from context
	currentUser, err := ctx.Get("current_user")
	if err != nil {
		return nil, err
	}

	// Create and fabricate presenters
	presenters, err := m.presenterFactory.Fabricate(
		members,
		map[string]interface{}{
			"current_user": currentUser,
			"presenter_class": "MembersPresenter",
		},
	)
	if err != nil {
		return nil, err
	}

	// Convert presenters to MemberPresenter type
	memberPresenters := make([]*service.MemberPresenter, len(presenters))
	for i, presenter := range presenters {
		if memberPresenter, ok := presenter.(*service.MemberPresenter); ok {
			memberPresenters[i] = memberPresenter
		}
	}

	return memberPresenters, nil
}

// preloadAssociations preloads all associations for the members
func (m *MembersPresentation) preloadAssociations(members []*service.Member) error {
	return m.membersPreloader.PreloadAll(members)
}

File: rust_app/src/controllers/concerns/members_presentation.rs
// Ported from: orig_app/app/controllers/concerns/members_presentation.rb (Ruby)
// Ported on: 2025-04-25
// This file implements the MembersPresentation concern in Rust.

use crate::models::user::User;
use crate::presenters::members_presenter::MembersPresenter;
use crate::services::members_preloader::MembersPreloader;
use crate::view::presenter::factory::PresenterFactory;

pub trait MembersPresentation {
    fn present_members(&self, members: Vec<User>) -> Vec<MembersPresenter>;
    fn preload_associations(&self, members: &[User]);
}

pub struct MembersPresentationImpl {
    current_user: User,
}

impl MembersPresentationImpl {
    pub fn new(current_user: User) -> Self {
        Self { current_user }
    }
}

impl MembersPresentation for MembersPresentationImpl {
    fn present_members(&self, members: Vec<User>) -> Vec<MembersPresenter> {
        // Preload associations for all members
        self.preload_associations(&members);

        // Create and return presenters for all members
        PresenterFactory::new(members, self.current_user.clone(), MembersPresenter::new).fabricate()
    }

    fn preload_associations(&self, members: &[User]) {
        MembersPreloader::new(members).preload_all();
    }
}
